### PR TITLE
Highlight only visible text

### PR DESCRIPTION
**Pull request by Moosems**
Here's the current plan:

1. Create a new tag named "MLCDS" to remember where Docstrings and Multi Line Comments (MLCs) start and end. This tag should not start with "Token" so that it is not deleted later.
2. Make a method that takes tokens, tags, and areas to remove and then returns the tokens and tags without those areas
3. Update the code to add MLCDS tags to the text widget where necessary. If a line starts or ends with a Docstring or MLC, add an MLCDS tag to the entire line.
4. When the visible area of the text widget changes, get the visible area, visible text, and line offset, as before.
5. Work with the MLCDS tags to add the necessary starts and ends to the visible text and then lex it using Pygments.
6. Splice the tags to remove parts that are not visible (columns to the left or right of the visible area) and remove any tags that are on emoji characters.
7. Add the remaining tags to the visible text and display it in the text widget.
In summary, the changes will make the text widget display only the visible area, while adding the necessary context outside the viewport using MLCDS tags.

### CODE REVIEW

**Comment by Moosems** [comment]
Could you review the plan and give your thoughts/ideas/suggestions/changes?

File: chlorophyll/codeview.py [hunk 1]
```diff
@@ -16,7 +16,12 @@
 
 color_schemes_dir = Path(__file__).parent / "colorschemes"
 
+def return_self(func):
```
Comment (rdbende) [inline chlorophyll/codeview.py:19]: Is this needed?

File: chlorophyll/codeview.py [hunk 2]
```diff
@@ -223,36 +193,46 @@ def configure(self, **kwargs) -> None:
...
+    def place(self, *args, **kwargs) -> CodeView:
         self._frame.place(*args, **kwargs)
 
-    def pack_forget(self) -> None:
+    @return_self
+    def pack_forget(self) -> CodeView:
         self._frame.pack_forget()
 
-    def grid_forget(self) -> None:
+    @return_self
```
Comment (rdbende) [inline chlorophyll/codeview.py:212]: Especially not here.

File: chlorophyll/codeview.py (hunk 1 above)
Comment (Moosems) [reply chlorophyll/codeview.py:19]: It makes life easier and more pythonic in most cases. It allows things like `cd = Codeview(root).pack()` which is something many new people do and if their life can be made easier, I find that better.

File: chlorophyll/codeview.py (hunk 2 above)
Comment (Moosems) [reply chlorophyll/codeview.py:212]: The pack, grid, place, and the forgets? That's practically the best best place for it imo.

File: chlorophyll/codeview.py (hunk 1 above)
Comment (rdbende) [reply chlorophyll/codeview.py:19]: I would say, more JavaScript-y.

Comment (Moosems) [reply chlorophyll/codeview.py:19]: Hmmm. So no go in your opinion?

File: chlorophyll/codeview.py (hunk 2 above)
Comment (Moosems) [reply chlorophyll/codeview.py:212]: Fair enough. I'll get rid of it in the next push then. With regards to the actual plan of action, do you have any thoughts (answer and converse in the PR :) )?

File: chlorophyll/codeview.py (hunk 1 above)
Comment (rdbende) [reply chlorophyll/codeview.py:19]: No. While I like fluent APIs, this is not the place to use it imo.

Comment (Moosems) [reply chlorophyll/codeview.py:19]: Will be removed in next push.

**Comment by rdbende** [comment]
I like your plan. The only thing I have concerns about is the emoji-in-tag-range
checking thingy. I'm not sure how would it perform.

**Comment by Moosems** [comment]
Check for a set of characters in the text, if any ranges are found, remove it from any relevant tags (might make a method that takes the tokens and tags and areas to remove and returns the tags without those areas.

**Comment by rdbende** [comment]
> Check for a set of characters in the text, if any ranges are found, remove it from any relevant tags (might make a method that takes the tokens and tags and areas to remove and returns the tags without those areas).

But that would run on every keypress?

**Comment by rdbende** [comment]
Oh no

**Comment by Moosems** [comment]
Yes, this runs on every keypress. Maybe I should add a check at the beginning of the method to check if anything has changed and if something has, only then will it all run.

**Comment by Moosems** [comment]
A method to scroll to where someone types that works properly is also in order.

**Comment by Moosems** [comment]
@rdbende, this runs *before* the char is typed. The check for MLCDS tags that might have been created comes *before* the highlighting though. The issue? The check for MLCDS tags relies on checking for `"Token.Literal.String.Doc", "Token.Comment.Multiline", "Token.Literal.String.Backtick"` Tokens. What would you recommend?

**Comment by Moosems** [comment]
I wonder if it's the lexing or tags that takes up the computation time. If it's just the tags then it wouldn't be that hard to simply add the tags that are in the visible area and lex the whole files such that MLCDS isn't an issue otherwise it's going to be the whole arduous process.
